### DISCUSSION TITLE
Blocking navigation (onBeforeNavigate?)

### DISCUSSION THREAD

**Discussion by astanb** [grahammendick/navigation#593] [Q&A] [answered]
I'd like to be able to block a user leaving a scene if there are unsaved changes. There is no mention of how to do this sort of thing in the docs that I can see, and from digging through the code I can see the onBeforeNavigate prop available when setting up states. But this seems like it's supposed to be set up when the app is initialised, whereas I'd like to be able to define the function that gets called from within the scene itself because that is where I will know what to do depending on the state of the scene.

Is there a preferred way of doing this?

I tried doing 

```

const {stateNavigator} =
    useContext<NavigationEvent<AppNavigation>>(NavigationContext);

  function canLeave() {
    return false;
  }

  useEffect(() => {
    stateNavigator.onBeforeNavigate(() => canLeave());
  }, [stateNavigator]);
```

But it causes strange glitchy behaviour where the scene is left but then it jumps back to it.

**Comment by grahammendick** [comment, marked as answer]
Yea, `onBeforeNavigate` blocks navigation but only if it happens in React first. It can't block navigations that happen on the native side first. So I'm guessing the 'glitchy behaviour' you're seeing is on iOS when you swipe/press back?

**Reply by astanb** [reply]
Yes that's right

**Reply by grahammendick** [reply]
Yea, that's where the navigation happens on native first. Then the navigation is sent to React to catch up. So you're blocking a navigation that's already happened

**Reply by astanb** [reply]
Is there anything that can be done about it? It seems like a pretty fundamental thing to be able to do.

**Reply by astanb** [reply]
FWIW React Navigation has this functionality https://reactnavigation.org/docs/preventing-going-back, and in the case of the native-stack explains how to handle it.

**Reply by grahammendick** [reply]
You could implement your own back button if they can't leave. This blocks back navigation. Then when they can leave you remove your button and the default back navigation will come back
```jsx
<NavigationBar>
  <LeftBar>
    {!canLeave() && <BarButton title="Back">}
  </LetfBar>
</NavigationBar>
```

**Reply by grahammendick** [reply]
That feature doesn't work on React Navigation if you're using the native stack - see the [limitations at the bottom of that page](https://reactnavigation.org/docs/preventing-going-back/)

**Reply by astanb** [reply]
I understand they don't support it natively but there is a workaround. Is there a way of disabling swipe to go back with this router? As your example above wouldn't avoid that.

**Reply by grahammendick** [reply]
Providing your own back button does disable the swipe

**Reply by astanb** [reply]
It does on iOS but doesn't seem to on Android

**Reply by astanb** [reply]
Ok using the react native `BackHandler` API seems to handle the Android case. I should be able to cobble something workable together then, thanks!

**Reply by grahammendick** [reply]
On Android, you can [implement your own BackHandler](https://reactnative.dev/docs/backhandler) or use `onBeforeNavigate`

**Reply by FluffyDiscord** [reply]
Is there universal option for Android and ios for back handler?

**Reply by grahammendick** [reply]
@FluffyDiscord Hey there, thanks for getting in touch, what is it you're trying to do?

**Reply by FluffyDiscord** [reply]
> @FluffyDiscord Hey there, thanks for getting in touch, what is it you're trying to do?

I would like to "catch" going back/pressing back on android or ios (swiping?) and conditionally either do something else (for example close modal in react native and deny going back) or do something before going back (but i guess i can use some different hook?)

**Reply by grahammendick** [reply]
On Android, you can [add your own BackHandler](https://reactnative.dev/docs/backhandler).
On iOS, there's no equivalent to the back handler. The swipe happens on the native side and there's no way to call back to the React side (synchronously). You can provide your own back button on iOS to prevent the swipe/back
```jsx
<NavigationBar>
  <LeftBar>
    {!canLeave() && <BarButton title="Back">}
  </LetfBar>
</NavigationBar>
```
